### Your name:

<pre> Manali Shinde </pre>

### Collaborators:

<pre> N/A</pre>


In [27]:
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

Open the housing data


In [28]:
# I was able to save the file on my home drive therefore did not need to use the path above 
housing = pd.read_csv("/Users/manalishinde/housing.csv")
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [29]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null float64
total_rooms           20640 non-null float64
total_bedrooms        20433 non-null float64
population            20640 non-null float64
households            20640 non-null float64
median_income         20640 non-null float64
median_house_value    20640 non-null float64
ocean_proximity       20640 non-null object
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [30]:
housing.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [31]:
housing["ocean_proximity"].value_counts()

<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: ocean_proximity, dtype: int64

### Build full pipeline for the data analysis following the example of the notebook.
 Hint: the main part requested to change is the algorithm used (Lasso regression)

If you want to learn more about the Lasso regression, see resources below:
- http://scikit-learn.org/stable/modules/linear_model.html#lasso
- https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-ridge-lasso-regression-python/

#### Considerations for building pipeline:

- Split data into training and testing sets below.
- Convert all categorical data to one-hot vectors below
- Normalize all non-categorical data 
-  Perform Lasso-based regression using a variety of values for $\alpha$ between 0 and 1 via a grid search where  *housing_labels* is the output and all other features are the input (similar to as seen in lecture two.)

In [32]:
housing.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
ocean_proximity        object
dtype: object

In [33]:
housing = pd.get_dummies(housing)
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  \
0       322.0       126.0         8.3252            452600.0   
1      2401.0      1138.0         8.3014            358500.0   
2       496.0       177.0         7.2574            352100.0   
3       558.0       219.0         5.6431            341300.0   
4       565.0       259.0         3.8462            342200.0   

   ocean_proximity_<1H OCEAN  ocean_proximity_INLAND  ocean_proximity_ISLAND  \
0                          0                       0                       0   
1                          0                       0                       0   
2                          0                       0                       0   
3                          0                       0                       0   
4                          0                       0                       0   

   ocean_proximity_NEAR BAY  ocean_proximity_NEAR OCEAN  
0                         1                           0  
1                         1                           0  
2                         1                           0  
3                         1                           0  
4                         1                           0

In [34]:
housing.dtypes

longitude                     float64
latitude                      float64
housing_median_age            float64
total_rooms                   float64
total_bedrooms                float64
population                    float64
households                    float64
median_income                 float64
median_house_value            float64
ocean_proximity_<1H OCEAN       uint8
ocean_proximity_INLAND          uint8
ocean_proximity_ISLAND          uint8
ocean_proximity_NEAR BAY        uint8
ocean_proximity_NEAR OCEAN      uint8
dtype: object

In [35]:
num_cols = housing.columns[housing.dtypes.apply(lambda c: np.issubdtype(c, np.number))]
num_cols
housing[num_cols].head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  \
0       322.0       126.0         8.3252            452600.0   
1      2401.0      1138.0         8.3014            358500.0   
2       496.0       177.0         7.2574            352100.0   
3       558.0       219.0         5.6431            341300.0   
4       565.0       259.0         3.8462            342200.0   

   ocean_proximity_<1H OCEAN  ocean_proximity_INLAND  ocean_proximity_ISLAND  \
0                          0                       0                       0   
1                          0                       0                       0   
2                          0                       0                       0   
3                          0                       0                       0   
4                          0                       0                       0   

   ocean_proximity_NEAR BAY  ocean_proximity_NEAR OCEAN  
0                         1                           0  
1                         1                           0  
2                         1                           0  
3                         1                           0  
4                         1                           0

In [36]:
from sklearn.preprocessing import Normalizer

In [37]:
housing= housing.dropna()

In [38]:
housing_pred = housing.drop("median_house_value", axis = 1)
housing_labels = housing.median_house_value
housing_pred.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'ocean_proximity_<1H OCEAN', 'ocean_proximity_INLAND',
       'ocean_proximity_ISLAND', 'ocean_proximity_NEAR BAY',
       'ocean_proximity_NEAR OCEAN'],
      dtype='object')

In [39]:
scaler = Normalizer().fit(housing_pred)

In [40]:
housing_norm = scaler.transform(housing_pred)

In [41]:
np.set_printoptions(precision=3)
print(housing_norm[0:5,:])

[[ -1.268e-01   3.931e-02   4.254e-02   9.131e-01   1.339e-01   3.341e-01
    1.307e-01   8.639e-03   0.000e+00   0.000e+00   0.000e+00   1.038e-03
    0.000e+00]
 [ -1.595e-02   4.942e-03   2.741e-03   9.266e-01   1.444e-01   3.134e-01
    1.485e-01   1.084e-03   0.000e+00   0.000e+00   0.000e+00   1.305e-04
    0.000e+00]
 [ -7.755e-02   2.401e-02   3.299e-02   9.307e-01   1.205e-01   3.147e-01
    1.123e-01   4.604e-03   0.000e+00   0.000e+00   0.000e+00   6.344e-04
    0.000e+00]
 [ -8.524e-02   2.639e-02   3.626e-02   8.883e-01   1.639e-01   3.891e-01
    1.527e-01   3.935e-03   0.000e+00   0.000e+00   0.000e+00   6.973e-04
    0.000e+00]
 [ -6.909e-02   2.139e-02   2.939e-02   9.195e-01   1.582e-01   3.193e-01
    1.464e-01   2.174e-03   0.000e+00   0.000e+00   0.000e+00   5.651e-04
    0.000e+00]]


In [42]:
from sklearn.model_selection import train_test_split
train_X, val_X, train_y, val_y = train_test_split(housing_norm, housing_labels,random_state = 0)

In [43]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
model = Lasso()

In [44]:
param_grid = [{'alpha':[1.0,0.0,0.1,0.01,0.001]}]
housing_grid = GridSearchCV(model, param_grid, cv=10)
housing_grid

GridSearchCV(cv=10, error_score='raise',
       estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid=[{'alpha': [1.0, 0.0, 0.1, 0.01, 0.001]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [45]:
housing_grid.fit(train_X, train_y)

GridSearchCV(cv=10, error_score='raise',
       estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid=[{'alpha': [1.0, 0.0, 0.1, 0.01, 0.001]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [46]:
housing_grid.grid_scores_

[mean: 0.31564, std: 0.02993, params: {'alpha': 1.0},
 mean: 0.31560, std: 0.03072, params: {'alpha': 0.0},
 mean: 0.31572, std: 0.03119, params: {'alpha': 0.1},
 mean: 0.31612, std: 0.03099, params: {'alpha': 0.01},
 mean: 0.31570, std: 0.03075, params: {'alpha': 0.001}]

In [47]:
housing_grid.best_estimator_

Lasso(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [48]:
test_scores = housing_grid.cv_results_
test_scores

{'mean_fit_time': array([ 0.299,  0.341,  0.27 ,  0.284,  0.305]),
 'mean_score_time': array([ 0.001,  0.001,  0.001,  0.001,  0.001]),
 'mean_test_score': array([ 0.316,  0.316,  0.316,  0.316,  0.316]),
 'mean_train_score': array([ 0.321,  0.323,  0.322,  0.323,  0.323]),
 'param_alpha': masked_array(data = [1.0 0.0 0.1 0.01 0.001],
              mask = [False False False False False],
        fill_value = ?),
 'params': [{'alpha': 1.0},
  {'alpha': 0.0},
  {'alpha': 0.1},
  {'alpha': 0.01},
  {'alpha': 0.001}],
 'rank_test_score': array([4, 5, 2, 1, 3], dtype=int32),
 'split0_test_score': array([ 0.281,  0.282,  0.281,  0.282,  0.282]),
 'split0_train_score': array([ 0.325,  0.326,  0.325,  0.326,  0.326]),
 'split1_test_score': array([ 0.366,  0.369,  0.369,  0.369,  0.369]),
 'split1_train_score': array([ 0.315,  0.317,  0.316,  0.317,  0.317]),
 'split2_test_score': array([ 0.285,  0.286,  0.283,  0.286,  0.286]),
 'split2_train_score': array([ 0.325,  0.326,  0.326,  0.326,  0.3

In [49]:
housing_value = model.fit(train_X, train_y)
housing_value

Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [50]:
housing_value.coef_

array([ -1039255.928,  -5051701.64 ,    157311.913,   -397771.019,
        -1151788.792,   -666934.018,    980415.959,   7644261.728,
         6606707.181, -50509785.005,         0.   ,         0.   ,
              -0.   ])

In [51]:
housing_value.predict(train_X)

array([ 103114.265,  218850.512,  139278.666, ...,  208518.704,
        118623.299,  134260.162])


Why is it necessary to normalize all continuous variables before performing Lasso? (OPTIONAL)

<pre>Since Lasso regression puts constraints on the size of the coefficients that you are using, it is important to  centre and reduce, and/or normalize the variables, thereby ensuring that they are all on the same scale. When you normalize, you ensure that the predictors do not outweight one another, making the model go one way or another  due to a particular predictor that has large numerical values (for example, the total bedroom feature at very   large numbers putting more weight on that feature than other ones). </pre>

### Conclusions
For what values of $\alpha$ does Lasso perform best? Does it perform as well on the housing data as the linear regressor from the lectures? Why do you think this is?

<pre> At alpha = 1 the Lasso perform best, with a standard deviation of 0.02993. Since we have a low standard        deviation, when performing a Lasso regression, it would be best to use alpha = 1 to get the most optimal features. It  does perform as well as the linear regressor from the textbook Chapter Two End to End Github code. This  is because we have normalized the numerical variables so that they perform well together. Also, since alpha = 1 performs best, this means that the coefficients are that of a simple linear regression analysis, producing      similar values.<pre> 

### Read appending B

- Reflect on your last data project, read appendix B. Then, write down a few of the checklist items that your last data project could have used. If you have not yet done a data project, then write down a few of the items that you found most interesting.


<pre> 
- performing a GridSearchCV is a good way to see which features are best and if the model you are using       performs well for this data 
- normalizing data ensure that the model is performing smoothly <pre> 

### Submit your notebook

Submit your solution to Blackboard
Make sure you rename your notebook to    
W2_UTORid.ipynb    
Example W2_adfasd01.ipynb
